skipchain/struct.go

```diff
@@ -1148,112 +1151,105 @@ func (db *SkipBlockDB) GetFuzzy(id string) (*SkipBlock, error) {
 		}
 
-		return err
+		// the search was supposed to go to the latest block.
+		if dest < 0 {
+			return nil
+		}
+		return xerrors.New("didn't find destination block")
 	})
 	return
 }
 
-// GetProofForID returns the shortest chain known from the genesis to the given
-// block using the highest forward-links available in the local db.
-func (db *SkipBlockDB) GetProofForID(bid SkipBlockID) (sbs Proof, err error) {
-	err = db.View(func(tx *bbolt.Tx) error {
-		target, err := db.getFromTx(tx, bid)
+// Iterate over all blocks until it's either the last block or the
+// one required by the call.
+func (db *SkipBlockDB) getPath(tx *bbolt.Tx, sbs Proof, links []*ForwardLink,
+	dest int) error {
+	sb := sbs[0]
+	for len(sb.ForwardLink) > 0 {
+		var fl *ForwardLink
+		var err error
+		fl, sb, err = db.getHighestJump(tx, sb, dest)
 		if err != nil {
-			return err
+			return xerrors.Errorf("while fetching next jump: %v", err)
 		}
-		if target == nil {
-			return errors.New("couldn't find the block")
+		links = append(links, fl)
+		sbs = append(sbs, sb)
+		if sb.Index == dest {
+			return nil
 		}
+	}
+	return nil
+}
 
-		sb, err := db.getFromTx(tx, target.SkipChainID())
-		if err != nil {
-			return err
+// Search for the closest block that is reachable before or at the destination.
+// dest < 0 indicates to chose the highest forward-link.
+func (db *SkipBlockDB) getHighestJump(tx *bbolt.Tx, start *SkipBlock,
+	dest int) (*ForwardLink, *SkipBlock, error) {
+	for i := len(start.ForwardLink) - 1; i >= 0; i-- {
+		// We can have holes in the forward links
+		if start.ForwardLink[i].IsEmpty() {
+			continue
 		}
-		if sb == nil {
-			// It should never happen if the previous is found.
-			return errors.New("couldn't find the genesis block")
+		sb, err := db.getFromTx(tx, start.ForwardLink[i].To)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("while fetching block from db: %v",
+				err)
 		}
 
-		sbs = append(sbs, sb)
-
-		for !sb.Hash.Equal(bid) && len(sb.ForwardLink) > 0 {
-			diff := math.Log(float64(target.Index - sb.Index))
-			base := math.Log(float64(sb.BaseHeight))
-			maxHeight := 0
-			if base != 0 {
-				maxHeight = int(math.Min(diff/base, float64(len(sb.ForwardLink)-1)))
+		if sb != nil {
+			if dest < 0 {
+				dest = sb.Index
 			}
-
-			id := sb.ForwardLink[maxHeight].To
-			sb, err = db.getFromTx(tx, id)
-			if err != nil {
-				return err
+			if sb.Index <= dest {
+				return start.ForwardLink[i], sb, nil
 			}
-
-			if sb == nil {
-				return errors.New("couldn't find one of the blocks")
-			}
-
-			if sb.Index <= sbs[len(sbs)-1].Index {
-				return ErrorInconsistentForwardLink
-			}
-
-			sbs = append(sbs, sb)
 		}
+	}
+	return nil, nil, xerrors.New("didn't find any appropriate block")
+}
 
-		return nil
-	})
+// GetProof returns the shortest chain from the genesis to the latest block
+// using the highest forward-links available in the local db
+func (db *SkipBlockDB) GetProof(sid SkipBlockID) (sbs []*SkipBlock,
```

Comment:
As it is pretty much the same as `GetFullProof`, maybe it can be remove?

Reply:
I'm hesitating - it is public, so theoretically it could be used elsewhere, and I didn't want to break it. But looking at the usage, it's only in the `skipchain` package, so it _should_ be fine. I can mark it `Deprecated` (even though we never _really_ removed anything, did we?)

Reply:
> I'm hesitating - it is public, so theoretically it could be used elsewhere, and I didn't want to break it. But looking at the usage, it's only in the `skipchain` package, so it _should_ be fine.

As you wish, I'm always in favor of reducing the number of exposed function, especially if it serves the same purpose.

> I can mark it `Deprecated` (even though we never _really_ removed anything, did we?)

Yeah, to be nice with semver, we should do a new major release where we remove the deprecated functions, but that might never happen.

Reply:
Renamed the methods to make more sense:
- GetProofFromIndex - takes a skipchain-ID and an index
- GetProof - for compatibility, deprecated, not used anymore in the code
- GetProofForLatest - calls `GetProofFromIndex` with magical value `-1` - replaces previous `GetProof`

And the return values of all these methods is now the same. The `Proof` structure got a `GetForwardLinks` method.